examples/hal/compute/main.rs

```diff
@@ -0,0 +1,162 @@
+    let shader = gpu.device.create_shader_module_from_source(include_bytes!("shader/collatz.comp"), pso::Stage::Compute).unwrap();
+
+    let (pipeline_layout, pipeline, set_layout, mut desc_pool) = {
+        let set_layout = gpu.device.create_descriptor_set_layout(&[
+                pso::DescriptorSetLayoutBinding {
+                    binding: 0,
+                    ty: pso::DescriptorType::StorageBuffer,
+                    count: 1,
+                    stage_flags: pso::STAGE_COMPUTE,
+                }
+            ],
+        );
+
+        let pipeline_layout = gpu.device.create_pipeline_layout(&[&set_layout]);
+        let entry_point = pso::EntryPoint { entry: "main", module: &shader };
+        let pipeline = gpu.device.create_compute_pipelines(&[(entry_point, &pipeline_layout)]).remove(0).expect("Error creating compute pipeline!");
+
+        let desc_pool = gpu.device.create_descriptor_pool(
+            1,
+            &[
+                pso::DescriptorRangeDesc {
+                    ty: pso::DescriptorType::StorageBuffer,
+                    count: 1,
+                },
+            ],
+        );
+        (pipeline_layout, pipeline, set_layout, desc_pool)
+    };
+
+    // For now this just panics if you didn't pass numbers. Could add proper error handling.
+    let numbers: Vec<u32> = std::env::args().skip(1).map(|s| u32::from_str(&s).unwrap()).collect();
+    let stride = std::mem::size_of::<u32>() as u64;
+
+    let (staging_memory, staging_buffer) = create_buffer(
+        &mut gpu, 
+        memory::CPU_VISIBLE | memory::COHERENT,
+        buffer::TRANSFER_SRC,
+        stride,
+        numbers.len() as u64
+    );
+
+    {
+        let mut writer = gpu.device.acquire_mapping_writer::<u32>(&staging_buffer, 0..stride * numbers.len() as u64).unwrap();
+        writer.copy_from_slice(&numbers);
+        gpu.device.release_mapping_writer(writer);
+    }
+
+    let (device_memory, device_buffer) = create_buffer(
+        &mut gpu,
+        memory::DEVICE_LOCAL,
+        buffer::TRANSFER_DST,
+        stride,
+        numbers.len() as u64
+    );
+
+    let desc_set = desc_pool.allocate_sets(&[&set_layout]).remove(0);
+    gpu.device.update_descriptor_sets(&[
+        pso::DescriptorSetWrite {
+            set: &desc_set,
+            binding: 0,
+            array_offset: 0,
+            write: pso::DescriptorWrite::StorageBuffer(vec![(&device_buffer, 0..stride * numbers.len() as u64)])
+        }
+    ]);
+
+    let mut command_pool = gpu.compute_queues[0].create_compute_pool(16, pool::CommandPoolCreateFlags::empty());
+    let fence = gpu.device.create_fence(false);
+    let submission = queue::Submission::new().submit(&[{
+        let mut command_buffer = command_pool.acquire_command_buffer();
+        command_buffer.copy_buffer(&staging_buffer, &device_buffer, &[command::BufferCopy { src: 0, dst: 0, size: stride * numbers.len() as u64}]);
+        command_buffer.bind_compute_pipeline(&pipeline);
```

Comment:
Missing pipeline barriers from TRANSFER to SHADER_WRITE|SHADER_READ for the SSBO/UAV and the other way around after the dispatch imo.

Reply:
Are these the only pipeline barriers necessary?